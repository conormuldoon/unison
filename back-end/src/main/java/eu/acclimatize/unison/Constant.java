package eu.acclimatize.unison;

public class Constant {
	
	public static final String LOCATION = "location";
	public static final String FORMAT = "d/M/yyyy";
	public static final String FROM_DATE = "fromDate";
	public static final String TO_DATE = "toDate";
	
	public static final String USERNAME = "username";
	public static final String PASSWORD = "password";
	
	public static final String PROPERTIES_FILE = "/application.properties";
	
	


}
