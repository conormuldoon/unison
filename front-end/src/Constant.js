
export const FORMAT = "D/M/YYYY";

export const API = "/apiacc";

export const LPER = "20\u1d57\u02b0 percentile";
export const HPER = "80\u1d57\u02b0 percentile";
export const MEDIAN = "Median";

export const LOW = "Low";
export const MEDIUM = "Medium";
export const HIGH = "High";
export const PRECIP = "Precipitation";
export const CL = "Cloud Level";
//export const API = "";

// Map coordinates
//export const CENTRE_LAT = 53.35014;
//export const CENTRE_LON = -6.266155;

// Oslo
export const CENTRE_LAT = 59.922326;
export const CENTRE_LON = 10.751560;
