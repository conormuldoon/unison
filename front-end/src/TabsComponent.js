import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import React, { useState,useEffect  } from 'react';
import "react-tabs/style/react-tabs.css";
import ChartComponent from './ChartComponent';

import {LPER,MEDIAN,HPER,LOW,MEDIUM,HIGH,CL} from './Constant';

import PropTypes from 'prop-types';





/**
 * A component that displays a number of tabs for ChartComponents for the phenomenom being tracked. 
 * For exmaple, for cloud level, there is a tab for high, medium, and low level clouds. Otherwise,
 * only a single ChartComponent is displayed.
 *  
 * @component
 */
function TabsComponent(props){

  const [tabs,setTabs]=useState(undefined);


  useEffect(()=>{

    function updateTabs(){
      if(props.minMax){
  
        setTabs([LPER,MEDIAN,HPER]);
    
      }else if(props.varCur===CL){
    
        setTabs([LOW,MEDIUM,HIGH]);
    
      }else{
        setTabs(undefined);
      }
    }

    if(props.data&&props.data.length>0){

        updateTabs(props.minMax,props.varCur===CL);     

      }else{
        
        setTabs(undefined);  
        
      }
    },[props.data,props.varCur,props.minMax]
  );
  

      function tabsPanel(index){
        return (<TabPanel >
          <div id="tabdiv" data-testid='chart' >
            <ChartComponent data={props.data} zoomDomain={props.zoomDomain} index={index}
              handleZoom={props.setZoomDomain}
              varCur={props.varCur} minMax={props.minMax}  />
          </div>
        </TabPanel>
        );
      }

      return (<div  id="chartdiv">
          {tabs&&<Tabs id="tabscomponent" defaultIndex={1}>
          <TabList >
            <Tab>{tabs[0]}</Tab>
            <Tab>{tabs[1]}</Tab>
            <Tab>{tabs[2]}</Tab>
          </TabList>

          {tabsPanel(0)}
          {tabsPanel(1)}
          {tabsPanel(2)}

        </Tabs>}

        {!tabs&&<div data-testid='chart' id="singlevar">
          <ChartComponent data={props.data} zoomDomain={props.zoomDomain} handleZoom={props.setZoomDomain}
            varCur={props.varCur} minMax={props.minMax} />
          </div>
        }
      </div>
    );

}


TabsComponent.propTypes ={
 
  /** An array of data to be displayed on one or more charts. */
  data: PropTypes.array,

  /** The current weather variable selected. */
  varCur: PropTypes.string,

  /** Determines wheather to display tabs for the precipitation weather variable. This is needed in that with
   * some locations and models only provide a single preciption value (no minimum or maximum).
   */
  minMax: PropTypes.bool,

  /** Determines the range of data to be displayed. */
  zoomDomain: PropTypes.object,

  /** A function used to change the zoom domain. */
  setZoomDomain: PropTypes.func,

  
}

export default TabsComponent;
