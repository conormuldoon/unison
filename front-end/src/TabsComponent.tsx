import PropTypes from 'prop-types';
import React from 'react';
import { Tab, TabList, TabPanel, Tabs } from 'react-tabs';
import "react-tabs/style/react-tabs.css";
import { CL, HIGH, HPER, LOW, LPER, MEDIAN, MEDIUM } from './Constant';

interface TabsProps {
  minMax: boolean;
  curVar: string;
  chartFactory: (index?: number) => React.ReactNode;
}
/**
 * A component that displays a number of tabs for ChartComponents for the phenomenom being tracked. 
 * For exmaple, for cloud level, there is a tab for high, medium, and low level clouds. Otherwise,
 * only a single ChartComponent is displayed.
 *  
 * @component
 */
function TabsComponent({ minMax, curVar, chartFactory }:
  TabsProps): JSX.Element {

  let tabs;

  if (minMax) {

    tabs = [LPER, MEDIAN, HPER];

  } else if (curVar === CL) {

    tabs = [LOW, MEDIUM, HIGH];

  }

  function tabsPanel(index: number) {
    const tChart = chartFactory(index);
    return (<TabPanel >
      <div id="tabdiv" data-testid='chart' >
        {tChart}
      </div>
    </TabPanel>
    );
  }

  let chart;
  if (!tabs) {
    chart = chartFactory()
  }

  return (<div id="chartdiv">
    {tabs && <Tabs id="tabscomponent" defaultIndex={1}>
      <TabList >
        <Tab>{tabs[0]}</Tab>
        <Tab>{tabs[1]}</Tab>
        <Tab>{tabs[2]}</Tab>
      </TabList>

      {tabsPanel(0)}
      {tabsPanel(1)}
      {tabsPanel(2)}

    </Tabs>}

    {!tabs && <div data-testid='chart' id="singlevar">
      {chart}
    </div>
    }
  </div>
  );

}


TabsComponent.propTypes = {

  chartFactory: PropTypes.func.isRequired,

  /** The current weather variable selected. */
  curVar: PropTypes.string.isRequired,

  /** Determines wheather to display tabs for the precipitation weather variable. This is needed in that with
   * some locations and models only provide a single preciption value (no minimum or maximum).
   */
  minMax: PropTypes.bool,


}

export default TabsComponent;
