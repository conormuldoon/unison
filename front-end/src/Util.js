
import {FORMAT} from './Constant';

import  {
  formatDate
} from 'react-day-picker/moment';

export const fromDate = () =>{

  const today=new Date(Date.now());
  const fDate=new Date(Date.now());
  fDate.setMonth(today.getMonth()-1);

  return formatDate(fDate,FORMAT);
}

export const today = () => {
  return formatDate(new Date(Date.now()), FORMAT);
};

const postObject = (body) =>{
  return {
    method: 'POST',
    headers: new Headers({
               'Content-Type': 'application/x-www-form-urlencoded',
      }),
    body: body
  }
}

const locCred = (location,uname,pword) =>{
  return 'location='+location+'&username='+uname+'&password='+pword;
}

export const removePostObject = (location,uname,pword) =>{
  return postObject(locCred(location,uname,pword));

}

export const locationPostObject = (location,uname,pword,lon,lat) =>{
  return postObject(locCred(location,uname,pword)+'&longitude='+lon+'&latitude='+lat);
}

export const varMapping = (varCur) =>{
  let variRequest=varCur.replace(/ /g,'');
  variRequest=variRequest.charAt(0).toLowerCase() + variRequest.slice(1);
  return variRequest;
}
