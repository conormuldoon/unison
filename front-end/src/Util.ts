
import { FORMAT } from './Constant';

import {
  formatDate
} from 'react-day-picker/moment';

import GeoJSON from 'geojson';

/**
 * Cretes a string for tomorrow's date.
 * 
 * @returns Tomorrow's date formatted in terms of day of the month, month, and year.
 */
export const tomorrow = (): string => {

  const now = Date.now();
  const today = new Date(now);
  const tomorrow = new Date(now);
  tomorrow.setDate(today.getDate() + 1)

  return formatDate(tomorrow, FORMAT);
}

/**
 * Cretes a string for tomorros's date.
 * 
 * @returns Tomorrow's date formatted in terms of day of the month, month, and year.
 */
export const today = (): string => {
  return formatDate(new Date(Date.now()), FORMAT);
};

/**
 * Converts a weather variable that may be selected by the user to a mapping suffix for an endpoint
 * on the server. The initial character of the weather variable is changed to lower case and words 
 * are concatenated. For example, 'Wind Direction' is converted to 'windDirection'. 
 * 
 * @param varCur The weather variable.
 * @returns The converted string.
 */
export const varMapping = (varCur: string): string => {
  let variRequest = varCur.replace(/ /g, '');
  variRequest = variRequest.charAt(0).toLowerCase() + variRequest.slice(1);
  return variRequest;
}

/**
 * Displays a message to te user that there was a problem connecting to the server.
 */
export const problemConnecting = (): void => {
  alert('There was a problem connecting to Unison. Please try again later.');
}


/**
 * Changes second word in a weather variable to lower case if present so that the second
 * word is not capitalised in a sentence. For example, the weather variable 'Wind Direction' 
 * would be changed to 'Wind direction'.
 * 
 * @param curVar The weather variable to be changed.
 * @returns The converted string.
 */
export const chartText = (curVar: string): string => {

  const vca = curVar.split(' ');

  let vc = vca[0];

  if (vca.length > 1) {
    vca[1] = vca[1].toLowerCase();

    vc += ' ' + vca[1];
  }

  return vc;
}

/**
 * Obtains the CSRF token with the key XSRF-TOKEN from the stored cookie.
 * 
 * @returns The CSRF toke.
 */
export function csrfToken(): string {

  // Modified regular expression from an anwswer to https://stackoverflow.com/questions/51109559/get-cookie-with-react
  // and removed unnecessary slash.
  return document.cookie.replace(/(?:(?:^|.*;\s*)XSRF-TOKEN\s*=\s*([^;]*).*$)|^.*$/, "$1");

}


/**
 * Creates request options for making put requests to the server for JSON. 
 * 
 * @param body The JSON body of the request.
 * @returns The options created. The optinos include the a X-XSRF-TOKEN header for the CSRF token.
 */
const putObject = (body: string) => {
  return {
    method: 'PUT',
    headers: new Headers({
      'Content-Type': 'application/geo+json',
      'X-XSRF-TOKEN': csrfToken()
    }),
    body: body
  }

}

/**
 * Creates put options with a GeoJSON representation of the location in the body.
 * 
 * @param locationName The name of the location to be put to the server.
 * @param longitude The longitute coordinate of the location. 
 * @param latitude The latitude coordinate of the location. 
 * 
 * @returns The put options create, which include a X-XSRF-TOKEN header for the CSRF token.
 */
export const locationPutObject = (locationName: string, longitude: string, latitude: string):
  Record<string, unknown> => {

  const location = { name: locationName, lng: longitude, lat: latitude };

  const geoJSONPoint = GeoJSON.parse(location, { Point: ['lat', 'lng'] });
  return putObject(JSON.stringify(geoJSONPoint));
}
