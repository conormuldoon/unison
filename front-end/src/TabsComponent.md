TabsComponent example:

```js
import {TODAY} from './Util';

import {useState} from 'react';
const [display,setDisplay] = useState(false);
function toggleDisplay(){
    if(display)setDisplay(false);
    else setDisplay(true);
}

if(display){
    window.scrollTo(0,0);
}

<div style={{"width":"80vw"}}>
    <button onClick={toggleDisplay}>{(display)?"Hide":"Display"}</button>
    
    {display&&<div id="popupdiv" >
        <button onClick={toggleDisplay}>Hide</button>
        <TabsComponent varCur="Precipitation" location="Oslo" fromDate={TODAY} toDate={TODAY}/>
        
    </div>}
</div>
```