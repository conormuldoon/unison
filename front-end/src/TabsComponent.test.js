import React from 'react';
import ReactDOM from 'react-dom';
import TabsComponent from './TabsComponent';
import fetchMock from 'fetch-mock';
import {FORMAT} from './Constant';
import { act } from 'react-dom/test-utils';
import "@testing-library/jest-dom/extend-expect";
import {varMapping} from './Util';
import  {
  formatDate
} from 'react-day-picker/moment';

import { render, waitForElement } from "react-testing-library"


const fromDate=formatDate(new Date(2018,11,20),FORMAT);
const toDate=formatDate(new Date(2019,3,4),FORMAT);
const tabsComponent=<TabsComponent varCur='Precipitation' location='UCD' fromDate={fromDate} toDate={toDate}/>;

const apiRequest='end:/precipitation?location=UCD&fromDate=20/12/2018&toDate=4/4/2019';
it('renders without crashing', async () => {


  fetchMock.get(apiRequest, []);

  const div = document.createElement('div');

  ReactDOM.render(tabsComponent,div);
  ReactDOM.unmountComponentAtNode(div);

  fetchMock.restore();
});

it('mathes snapshot', () => {

  fetchMock.get(apiRequest, []);

  const {container} = render(tabsComponent);

  expect(container).toMatchSnapshot();

  fetchMock.restore();

});

it("doesn't crash when no data obtained from server", async ()=>{
  const date='2/4/2019';
  const location='UCD';
  fetchMock.get('end:/humidity?location='+location+'&fromDate='+date+'&toDate='+date,[]);
  const {getByTestId} = render(<TabsComponent varCur='Humidity' location={location} fromDate={date} toDate={date} />);

  const chartDiv=await waitForElement(()=>getByTestId('chart'));
  fetchMock.restore();

});

const addChart = async (weatherVariable,yLabel,data) =>{
  const date='2/4/2019';
  const location='UCD';

  fetchMock.get('end:/'+varMapping(weatherVariable)+'?location='+location+'&fromDate='+date+'&toDate='+date,
      data);

  const {getByText} = render(<TabsComponent varCur={weatherVariable} location={location} fromDate={date} toDate={date}  />);

  await waitForElement(()=>getByText(yLabel));

  fetchMock.restore();

}

it('adds a chart for temperature', async () =>{
  await addChart('Temperature','Celsius',[
        {
           "date": "2019-04-01T23:00:00.000+0000",
           "temperature": 5.5
       },
       {
           "date": "2019-04-02T00:00:00.000+0000",
           "temperature": 4.8
       },
       {
           "date": "2019-04-02T01:00:00.000+0000",
           "temperature": 4.5
       },
       {
           "date": "2019-04-02T02:00:00.000+0000",
           "temperature": 3.9
       },
       {
           "date": "2019-04-02T03:00:00.000+0000",
           "temperature": 3.7
       },
       {
           "date": "2019-04-02T04:00:00.000+0000",
           "temperature": 3.4
       },
       {
           "date": "2019-04-02T05:00:00.000+0000",
           "temperature": 3.2
       },
       {
           "date": "2019-04-02T06:00:00.000+0000",
           "temperature": 3.4
       },
       {
           "date": "2019-04-02T07:00:00.000+0000",
           "temperature": 3.9
       },
       {
           "date": "2019-04-02T08:00:00.000+0000",
           "temperature": 4.9
       },
       {
           "date": "2019-04-02T09:00:00.000+0000",
           "temperature": 6.1
       },
       {
           "date": "2019-04-02T10:00:00.000+0000",
           "temperature": 6.1
       },
       {
           "date": "2019-04-02T11:00:00.000+0000",
           "temperature": 7.5
       },
       {
           "date": "2019-04-02T12:00:00.000+0000",
           "temperature": 8.9
       }
  ]);
});


it('adds tabs for precipitation', async () =>{

  await addChart('Precipitation','Millimetres', [{
        "date": "2019-04-01T23:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0.1
        }
    },
    {
        "date": "2019-04-02T00:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T01:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T02:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T03:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T04:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T05:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T06:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T07:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T08:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T09:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    },
    {
        "date": "2019-04-02T10:00:00.000+0000",
        "precipitation": {
            "value": 0,
            "minvalue": 0,
            "maxvalue": 0
        }
    }]);



});
