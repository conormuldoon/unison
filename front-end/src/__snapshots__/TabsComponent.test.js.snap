// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathes snapshot 1`] = `
<div>
  <div
    data-testid="chart"
    id="chartdiv"
  >
    <div
      data-testid="chart"
      id="singlevar"
    >
      <div
        class="VictoryContainer"
        style="height: 100%; width: 100%; user-select: none; pointer-events: none; position: relative;"
      >
        <svg
          aria-labelledby="victory-container-2-title victory-container-2-desc"
          height="350"
          role="img"
          style="pointer-events: all; width: 100%; height: 100%;"
          viewBox="0 0 350 350"
          width="350"
        >
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #90A4AE; fill: transparent; stroke-width: 2; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="50"
              x2="300"
              y1="300"
              y2="300"
            />
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="50"
                y1="300"
                y2="50"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="50"
                y1="300"
                y2="305"
              />
              <text
                direction="inherit"
                dx="0"
                dy="10.26"
                x="50"
                y="313"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="middle"
                  x="50"
                >
                  01 AM
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="300"
                x2="300"
                y1="300"
                y2="50"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="300"
                x2="300"
                y1="300"
                y2="305"
              />
              <text
                direction="inherit"
                dx="0"
                dy="10.26"
                x="300"
                y="313"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="middle"
                  x="300"
                >
                  .001
                </tspan>
              </text>
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: #90A4AE; fill: transparent; stroke-width: 2; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="50"
              x2="50"
              y1="50"
              y2="300"
            />
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="300"
                y1="250"
                y2="250"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="45"
                y1="250"
                y2="250"
              />
              <text
                direction="inherit"
                dx="0"
                dy="4.26"
                x="37"
                y="250"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="end"
                  x="37"
                >
                  0.2
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="300"
                y1="200"
                y2="200"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="45"
                y1="200"
                y2="200"
              />
              <text
                direction="inherit"
                dx="0"
                dy="4.26"
                x="37"
                y="200"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="end"
                  x="37"
                >
                  0.4
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="300"
                y1="150"
                y2="150"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="45"
                y1="150"
                y2="150"
              />
              <text
                direction="inherit"
                dx="0"
                dy="4.26"
                x="37"
                y="150"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="end"
                  x="37"
                >
                  0.6
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="300"
                y1="100"
                y2="100"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="45"
                y1="100"
                y2="100"
              />
              <text
                direction="inherit"
                dx="0"
                dy="4.26"
                x="37"
                y="100"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="end"
                  x="37"
                >
                  0.8
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #ECEFF1; fill: none; stroke-dasharray: 10, 5; stroke-linecap: round; stroke-linejoin: round; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="300"
                y1="50"
                y2="50"
              />
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #90A4AE; fill: transparent; size: 5px; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
                vector-effect="non-scaling-stroke"
                x1="50"
                x2="45"
                y1="50"
                y2="50"
              />
              <text
                direction="inherit"
                dx="0"
                dy="4.26"
                x="37"
                y="50"
              >
                <tspan
                  dx="0"
                  style="font-family: 'Roboto', 'Helvetica Neue', Helvetica, sans-serif; font-size: 12px; letter-spacing: normal; padding: 8px; fill: #455A64; stroke: transparent; stroke-width: 0;"
                  text-anchor="end"
                  x="37"
                >
                  1.0
                </tspan>
              </text>
            </g>
          </g>
        </svg>
        <div
          style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
        >
          <svg
            height="350"
            style="overflow: visible; width: 100%; height: 100%;"
            viewBox="0 0 350 350"
            width="350"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
